libs/unity/library/Runtime/Scripts/Media/MicrophoneSource.cs

```diff
@@ -36,77 +32,55 @@ public class MicrophoneSource : AudioTrackSource
         protected bool _autoGainControl = true;
 
 #if !UNITY_EDITOR && UNITY_ANDROID
-        protected bool _androidRecordAudioRequestPending = false;
+        protected TaskCompletionSource<bool> _androidPermissionRequestTcs;
         protected float _androidRecordAudioRequestRetryUntilTime = 0f;
 #endif
 
-        protected async void OnEnable()
+        private AsyncInitHelper<WebRTC.AudioTrackSource> _initHelper = new AsyncInitHelper<WebRTC.AudioTrackSource>();
+
+        protected void OnEnable()
         {
-            if (Source != null)
-            {
-                return;
-            }
+            Debug.Assert(Source == null);
+            _initHelper.TrackInitTask(InitAsync());
+        }
 
-#if !UNITY_EDITOR && UNITY_ANDROID
-            // Ensure Android binding is initialized before accessing the native implementation
-            Android.Initialize();
+        protected async Task<WebRTC.AudioTrackSource> InitAsync()
+        {
 
-            // Check for permission to access the camera
-            if (!Permission.HasUserAuthorizedPermission(Permission.Microphone))
-            {
-                if (!_androidRecordAudioRequestPending)
-                {
-                    // Monitor the OnApplicationFocus(true) event during the next 5 minutes,
-                    // and check for permission again each time (see below why).
-                    _androidRecordAudioRequestPending = true;
-                    _androidRecordAudioRequestRetryUntilTime = Time.time + 300;
-
-                    // Display dialog requesting user permission. This will return immediately,
-                    // and unfortunately there's no good way to tell when this completes. As a rule
-                    // of thumb, application should lose focus, so check when focus resumes should
-                    // be sufficient without having to poll every frame.
-                    Permission.RequestUserPermission(Permission.Microphone);
-                }
-                return;
-            }
-#endif
 
-#if UNITY_WSA && !UNITY_EDITOR
-            // Request access to audio capture. The OS may show some popup dialog to the
-            // user to request permission. This will succeed only if the user approves it.
-            try
+            // Continue the task outside the Unity app context, in order to avoid deadlock
+            // if OnDisable waits on this task.
+            bool accessGranted = await RequestAccessAsync().ConfigureAwait(false);
```

Comment:
This is not what you are doing; you are allowing the task to run on any thread, you are not forbidding it to run on the Unity app thread. AFAIK the scheduler is allowed to decide to run on the same thread even with `ConfigureAwait(false)`; it's up to it to decide between this thread and another.

Reply:
Discussed in chat. According to https://devblogs.microsoft.com/dotnet/configureawait-faq/, `ConfigureAwait(false)` prevents the continuation from being scheduled on the current SynchronizationContext, thus effectively pushing it into the default thread pool. We agree that unless some unlikely circumstances (Unity app thread belongs to the default thread pool; thread pool has no available threads) are true there should be no deadlock.